psydac/linalg/utilities.py

```diff
@@ -32,4 +44,46 @@ def array_to_stencil(x, Xh):
     else:
         raise ValueError('Xh must be a StencilVectorSpace or a BlockVectorSpace')
 
+    u.update_ghost_regions()
+    return u
+
+def petsc_to_stencil(x, Xh):
+    """ converts a numpy array to StencilVector or BlockVector format"""
+    x = x.array
+    if isinstance(Xh, BlockVectorSpace):
+        u = BlockVector(Xh)
+        if isinstance(Xh.spaces[0], BlockVectorSpace):
+            for d in range(len(Xh.spaces)):
+                starts = [np.array(V.starts) for V in Xh.spaces[d].spaces]
+                ends   = [np.array(V.ends)   for V in Xh.spaces[d].spaces]
+
+                for i in range(len(starts)):
+                    g = tuple(slice(s,e+1) for s,e in zip(starts[i], ends[i]))
+                    shape = tuple(ends[i]-starts[i]+1)
+                    u[d][i][g] = x[:np.product(shape)].reshape(shape)
+                    x       = x[np.product(shape):]
+
+        else:
+            starts = [np.array(V.starts) for V in Xh.spaces]
+            ends   = [np.array(V.ends)   for V in Xh.spaces]
+
+            for i in range(len(starts)):
+                g = tuple(slice(s,e+1) for s,e in zip(starts[i], ends[i]))
+                shape = tuple(ends[i]-starts[i]+1)
+                u[i][g] = x[:np.product(shape)].reshape(shape)
+                x       = x[np.product(shape):]
+
+    elif isinstance(Xh, StencilVectorSpace):
+
+        u =  StencilVector(Xh)
+        starts = np.array(Xh.starts)
+        ends   = np.array(Xh.ends)
+
+        g = tuple(slice(s,e+1) for s,e in zip(starts, ends))
+        shape = tuple(ends-starts+1)
+        u[g] = x[:np.product(shape)].reshape(shape)
+    else:
+        raise ValueError('Xh must be a StencilVectorSpace or a BlockVectorSpace')
+
+    u.update_ghost_regions()
     return u
```

Comment:
I think that you can write this function as:
```python
def petsc_to_stencil(x, Xh):
    assert isinstance(x, ...)
    return array_to_stencil(x.array, Xh)
```
So why duplicating so much code?

Reply:
ah yes, you are right